common/page.go

```diff
@@ -364,6 +363,127 @@ func (p *Page) initEvents() {
 	}()
 }
 
+func (p *Page) urlGroupingName(ctx context.Context, urlTag string) (string, bool) {
+	p.eventHandlersMu.RLock()
+
+	// If there are no handlers for EventConsoleAPICalled.
+	if _, ok := p.eventHandlers[EventPageMetricCalled]; !ok {
+		p.eventHandlersMu.RUnlock()
+
+		return "", false
+	}
+	p.eventHandlersMu.RUnlock()
+
+	var name string
+	var nameChanged bool
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	p.eventHandlersMu.RLock()
+
+	for _, h := range p.eventHandlers[EventPageMetricCalled] {
+		// A handler can register another handler from within itself. This is
+		// the reason to unlock the mutex before calling the handler.
+		p.eventHandlersMu.RUnlock()
+
+		nameCh := make(chan string)
+		em := &ExportedMetric{
+			ctx:    ctx,
+			urlTag: urlTag,
+			nameCh: nameCh,
+		}
+
+		// handler runs on another goroutine.
+		h(em)
+
+		// We want to wait for the name to come in from the handler. When the
+		// nameCh is closed by the handler the name is returned back upstream.
+		ok := true
+		for ok {
+			select {
+			case n, o := <-nameCh:
+				ok = o
+				if ok {
+					name = n
+					nameChanged = true
+				}
+			case <-ctx.Done():
+				return "", false
+			}
+		}
+
+		p.eventHandlersMu.RLock()
+	}
+	p.eventHandlersMu.RUnlock()
+
+	p.logger.Debugf("URLGroupingName", "name: %q nameChanged: %v", name, nameChanged)
+
+	return name, nameChanged
+}
+
+// ExportedMetric is the type that is exported to JS. It is currently only used to
+// match on the urlTag and return a name when a match is found.
+type ExportedMetric struct {
+	ctx context.Context
+
+	// The url value from the metric's url tag. It will be used to match
+	// against the url grouping regexs.
+	urlTag string
+
+	// nameCh is used to return the url tag metric name when a match is found.
+	// It is also used to help sync the call to the handler and the caller.
+	// We need to wait for the handler to complete before proceeding to return
+	// the name of the url grouping.
+	nameCh chan<- string
+}
+
+// URLGroups will contain all the url groupings.
+type URLGroups struct {
+	Groups []URLGroup
+}
+
+// URLGroup contains the single url regex and the name to give to the metric
+// if a match is found.
+type URLGroup struct {
+	// This is a regex.
+	URL string
+	// The name to send back to the caller of the handler.
+	Name string
+}
+
+// Completed will allow the caller of the handler to know that the handler
+// has completed and therefore to continue the flow.
+func (e *ExportedMetric) Completed() {
+	close(e.nameCh)
+}
+
+// GroupURLTag will find a match given the URLGroups and the URL from the metric
+// tag and send the name via a channel back to the caller of the handler.
+func (e *ExportedMetric) GroupURLTag(callBack func(pattern, url string) (bool, error), groups URLGroups) error {
+	for _, g := range groups.Groups {
+		if g.Name == "" {
```

Comment:
Do we need `strings.Trim` here as users can specify `"   "` as a name?

Reply:
Good spot. Fixed in 14f9fd9302c7638948f64ecc1c5cbfb0d30abe77